pkg/reconciler/service/reconciler.go

```diff
@@ -33,23 +33,35 @@ const (
 	defaultRetryDelay    = 30 * time.Second
 	defaultTimeout       = 10 * time.Minute
 	defaultWorkers       = 100
+	defaultWorkspace     = "."
+)
+
+var (
+	wsFactory *workspace.Factory //singleton
+	m         sync.Mutex
 )
 
+type ActionContext struct {
+	KubeClient       kubernetes.Client
+	WorkspaceFactory *workspace.Factory
+	Context          context.Context
```

Comment:
I know why you set Context here, but I found this article of creators golang - [context-and-structs](https://blog.golang.org/context-and-structs)

Quote:
Contexts should not be stored inside a struct type, but instead passed to each function that needs it.


It's only suggestion, we can stay with current solution.